### PR TITLE
Filter positive items for ranking evaluation

### PR DESCRIPTION
**Pull request by tqtg**
### Description

Previously, we did not filter positive items in training when doing evaluation on test/validation set for most of the ranking metrics. As a result, it harms performance of some models that indeed rank those items to the top. In this PR, we first explicitly filter out those items when calling `Recommender.rank()` function using `item_indices`, we then modify all ranking metrics to obey this logic. To summarize, positive items in training will not be involved during evaluation.

How are the ranking metrics affected?
- `AUC`: not affected - previous logic already took it into account.
- `Precision`: affected - expected the results to be increased.
- `Recall`: affected - expected the results to be increased.
- `NDCG`: affected - expected the results to be increased.
- `NCRR`: affected - expected the results to be increased.
- `MAP`: affected - expected the results to be increased.
- `MRR`: affected - expected the results to be increased.

How are the rating metrics affected?
No. This change only affects ranking metrics.

### Related Issues

#503

### Checklist:

- [x] I have added tests.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated `README.md` (if you are adding a new model).
- [ ] I have updated `examples/README.md` (if you are adding a new example).
- [ ] I have updated `datasets/README.md` (if you are adding a new dataset).

### CODE REVIEW

**Comment by tqtg** [comment]
Hi @amirj, we're making some changes to how the ranking metrics being computed, and these changes will break the logic of [PropensityStratifiedEvaluation](https://github.com/PreferredAI/cornac/blob/67c28bca66db4a7aba31e3e51c5a0f66d3b7f42d/cornac/eval_methods/propensity_stratified_evaluation.py#L130C50-L130C50). Could you please have a look and see if we could mitigate this with some modification to the eval method?

**Comment by tqtg** [comment]
@saghiles @lthoang please help spend some time on this PR and see if this makes sense to you. The best way is to write a few test cases. Let me know if you have any comments

File: cornac/eval_methods/base_method.py [hunk 1]
```diff
@@ -187,18 +188,30 @@ def pos_items(csr_row):
...
+        u_gt_neg_mask[test_pos_items + val_pos_items + train_pos_items] = 0
+
+        # filter items being considered for evaluation
+        if exclude_unknowns:
+            item_indices = np.arange(train_set.num_items)
+            u_gt_pos_mask = u_gt_pos_mask[: train_set.num_items]
+            u_gt_neg_mask = u_gt_neg_mask[: train_set.num_items]
+        else:
+            item_indices = np.arange(test_set.num_items)
+        item_indices = np.nonzero(u_gt_pos_mask + u_gt_neg_mask)[0]
```
Comment (saghiles) [inline cornac/eval_methods/base_method.py:202]: Given this line, do we still need item_indices inside the above if - else block?

File: cornac/metrics/ranking.py [hunk 2]
```diff
@@ -117,7 +120,7 @@ def compute(self, gt_pos, pd_rank, **kwargs):
 
         """
         dcg = self.dcg_score(gt_pos, pd_rank, self.k)
-        idcg = self.dcg_score(gt_pos, np.argsort(gt_pos)[::-1], self.k)
+        idcg = self.dcg_score(gt_pos, np.concatenate([gt_pos, gt_neg]), self.k)
```
Comment (saghiles) [inline cornac/metrics/ranking.py:123]: Do we need to concatenate gt_pos with gt_neg here? Wouldn't it be enough if we just call  self.dcg_score(gt_pos, gt_pos, self.k)

**Review by saghiles — APPROVED** [review]
@tqtg, I left two comments please have a look. The rest looks good to me.

**Comment by lthoang** [comment]
@tqtg The two comments of @saghiles make sense to me. The rests LGTM.

**Comment by tqtg** [comment]
@saghiles @lthoang thanks both for the comments which are now fixed. If no further suggestions, I'll go ahead and merge this PR.

**Pull request merged**